### ISSUE TITLE
Cell containing Vaadin charts disappear when scrolling grid

### ISSUE THREAD

**Issue by ChrisKClark** [datenhahn/componentrenderer#25]
Chart components lose their size & visibility when their rows are not immediately visible in the grid. Any event that triggers a server request seems to cause these rows to become visible. Other component types (example combobox) do not seem to have this issue. See the attached image for what I'm trying to describe.
![twocomponents](https://cloud.githubusercontent.com/assets/20171716/16389401/06543aaa-3c5b-11e6-9cb9-8601f5ddcc93.PNG)

Steps to reproduce using the provided code example: 
1. Load the page; you should see a grid that includes a "Chart" column which contains sparklines. 
2. Scroll the grid down a few pages, passing the section where blanks are seen instead of the sparklines. 
3. Click one row in the Baz, Foo, or Bar column. 
4. Scroll up half a page, noting the blank sparklines. (See attached BlankRows.png) 
5. Scrolling further or clicking a row will cause the Sparklines in visible rows to render properly.

If you examine the attached "HiddenChartBeforeClickEvent.png" and compare against the attached "HiddenChartAfterClickEvent.png", I have underlined in the "After" image the markup attributes which are modified when the user clicks a row, thus causing the sparklines to become visible. Note the difference in the height & width attributes (changing from 0 to values >0) and the visibility attributes (changing from "hidden" to "visible").
![hiddenchart_beforeclickevent](https://cloud.githubusercontent.com/assets/20171716/16389618/f53d88f6-3c5b-11e6-9892-8b5a1f55085b.PNG)
![hiddenchart_afterclickevent](https://cloud.githubusercontent.com/assets/20171716/16389622/faa2a984-3c5b-11e6-81f0-75076edbabac.PNG)

Code used in a basic project UI class:

`protected void init(VaadinRequest vaadinRequest) {
        final VerticalLayout layout = new VerticalLayout();
        layout.addComponents(createGrid());
        layout.setMargin(true);
        layout.setSpacing(true);

```
    setContent(layout);
}

private Component createGrid() {
    BeanItemContainer<MyPojo> container = new BeanItemContainer<MyPojo>(MyPojo.class);
    for (int i = 0; i < 1000; ++i) {
        container.addBean(new MyPojo(i));
    }

    GeneratedPropertyContainer gpc = new GeneratedPropertyContainer(container);
    Grid grid = new Grid(gpc);
    grid.setWidth(900, Unit.PIXELS);
    grid.setSelectionMode(SelectionMode.MULTI);
    gpc.addGeneratedProperty("chart", new ComponentPropertyGenerator<>(MyPojo.class, pojo -> {
        return new Sparkline(130, 30, pojo.getNumbers());
    }));
    gpc.addGeneratedProperty("combo", new ComponentPropertyGenerator<>(MyPojo.class, pojo -> {
        return new ComboBox("Test", Arrays.asList(pojo.getNumbers()));
    }));
    grid.getColumn("chart").setRenderer(new ComponentRenderer());
    grid.getColumn("combo").setRenderer(new ComponentRenderer());
    grid.setColumns("baz", "foo", "bar", "chart", "combo");
    grid.addItemClickListener(
            event -> grid.getColumn("chart").setHeaderCaption("Chart" + String.valueOf(++clickCount)));
    return grid;
}


public class MyPojo {
    String foo = "foo";
    String bar = "bar";
    int baz = -1;
    final Number[] numbers;
    final Random rand;

    public MyPojo(int i) {
        baz = i;
        numbers = new Number[20];
        rand = new Random(baz);
    }

    public Number[] getNumbers() {
        IntStream.range(0, 20).forEach(index -> numbers[index] = rand.nextInt());
        return numbers;
    }

    public String getFoo() {
        return foo;
    }

    public void setFoo(String foo) {
        this.foo = foo;
    }

    public String getBar() {
        return bar;
    }

    public void setBar(String bar) {
        this.bar = bar;
    }

    public int getBaz() {
        return baz;
    }
}`
```

**Comment by datenhahn** [comment]
Could be reproduced in Chrome but not in Firefox, investigating

**Comment by ChrisKClark** [comment]
Sorry, I should have provided browser info. I have tried this in Chrome, IE 11 and Firefox 38.6.0 ESR, and the issue is consistently reproducible in each.

**Comment by datenhahn** [comment]
I think I traced it down to the source. If I assume correctly it has to do with how Highcharts/Highstock works and how the vaadin grid optimizes rendering. 

_guessing_ (because I don't want to debug minimized Highcharts code)
The chart tries to update its size depending on its surrounding elements, probably also to prevent rendering expensive charts which are not even visible. Somehow it gets the information that either the element surrounding it is 0 size or it is not visible and therefore sets its own size to 0. I think that happens because it measures something (size/visibility?) before the grid has finished rendering/setting its sizes.
_guessing_

I tried something which seems to work, but I have to investigate more. If I detect a HighchartWidget I schedule a call to the updateSize method after the browser's event loop is finished using the GWT scheduler. That seems to redraw the chart with the correct sizes. I could not reproduce the bug when using this code. A problem with that specific code is, that it requires the vaadin-charts dependency which I don't want to pull into the component renderer. I am thinking about using reflection/native JS to get around the dependency.

``` java

    @Override
    public void render(RendererCellReference rendererCellReference, ComponentConnector componentConnector,
                       SimplePanel panel)
    {
        if (componentConnector != null) {
            panel.setWidget(componentConnector.getWidget());
            if(componentConnector.getWidget() instanceof HighchartWidget) {
                final ComponentConnector passConnector = componentConnector;
                Scheduler.get().scheduleDeferred(new Scheduler.ScheduledCommand() {
                @Override
                public void execute() {
                    HighchartWidget chart = (HighchartWidget) passConnector.getWidget();
                    chart.updateSize();
                }
            });

            }
        } else {
            panel.clear();
        }
    }
```

Also I think I found a new bug. During the initial render the grid's automatic column resizing also seems to feed too small values to Highcharts (because it renders too early) resulting in the first few rows being renderered too small. One scroll up/down fixes that (because then the grid's columns are already right sized). I think of a similar fix as the one for the vanishing charts.

So bottom-line: nasty, but probably fixable.

If you have time please checkout the branch "chartfix" (clone my repo and do "git checkout chartfix")  containing the fix and build it with:
     mvn clean install

Then the version "1.0.0-chartfix" should be available in your repo. Use that in your project, **rebuild the widgetset** and check if the reported bug disappears with that fix.

I need a couple of more days to finalize the fix and tap into the grid auto-column resize mechanism for the initial updateSize-call after grid resizes its columns.

**Comment by ChrisKClark** [comment]
I noticed the same squishy-charts issue, and the workaround I used was to set the dimensions explicitly via the chart configuration object:

`
 Sparkline sparkline = new Sparkline(130, 30, pojo.getNumbers());
 sparkline.getConfiguration().getChart().setWidth(130);
 sparkline.getConfiguration().getChart().setHeight(30);       
`

I will checkout your branch and give the fix a try. Thank you for looking into it so quickly. I'll let you know how it goes.

**Comment by ChrisKClark** [comment]
Confirmed: 1.0.0-chartfix fixes the issue for me too.

**Comment by datenhahn** [comment]
Ok great, I just pushed a draft fix for the initial load "chart too small" problem. For testing purposes I just added into an existing extension, but I will pull it out into a separate one, if you also had that problem (initial load shows first row charts too small) then you can pull the branch again and rebuild the renderer + wigetset and try that as well. You have to add this to your code after you instantiated the grid:
        DetailsKeysExtension.extend(grid);

It will add this handler to the grid:

``` java
        final GridConnector gridConnector = ((GridConnector) target);
        final Timer timer = new Timer() {
            @Override
            public void run() {
                for(ServerConnector connector : gridConnector.getChildren()) {
                    if(connector instanceof ChartConnector) {
                        ((HighchartWidget) ((ChartConnector) connector).getWidget()).updateSize();
                    }
                }
            }
        };
        grid.addDataAvailableHandler(new DataAvailableHandler() {
            @Override
            public void onDataAvailable(DataAvailableEvent event) {
                Range range = event.getAvailableRows();
                if(range.contains(0)) {
                    timer.schedule(50);
                }
            }

        });
```

**Comment by datenhahn** [comment]
Hope that helps you to continue development for now until I release the final solution

**Comment by ChrisKClark** [comment]
It does the trick. Thanks again!

**Comment by datenhahn** [comment]
Released version 1.0.1 containing the fix:

```
<dependency>
   <groupId>de.datenhahn.vaadin</groupId>
   <artifactId>componentrenderer</artifactId>
   <version>1.0.1</version>
</dependency>
```
